File: library/src/main/java/com/yarolegovich/slidingrootnav/SlideGravity.java
package com.yarolegovich.slidingrootnav;

import android.support.v4.widget.ViewDragHelper;

/**
 * Created by yarolegovich on 25.03.2017.
 */

public enum SlideGravity {

    LEFT {
        @Override
        Helper createHelper() {
            return new LeftHelper();
        }
    },
    RIGHT {
        @Override
        Helper createHelper() {
            return new RightHelper();
        }
    };

    abstract Helper createHelper();

    interface Helper {

        int getLeftAfterFling(float flingVelocity, int maxDrag);

        int getLeftToSettle(float dragProgress, int maxDrag);

        int getRootLeft(float dragProgress, int maxDrag);

        float getDragProgress(int viewLeft, int maxDrag);

        int clampViewPosition(int left, int maxDrag);

        void enableEdgeTrackingOn(ViewDragHelper dragHelper);
    }

    static class LeftHelper implements Helper {

        @Override
        public int getLeftAfterFling(float flingVelocity, int maxDrag) {
            return flingVelocity > 0 ? maxDrag : 0;
        }

        @Override
        public int getLeftToSettle(float dragProgress, int maxDrag) {
            return dragProgress > 0.5f ? maxDrag : 0;
        }

        @Override
        public int getRootLeft(float dragProgress, int maxDrag) {
            return (int) (dragProgress * maxDrag);
        }

        @Override
        public float getDragProgress(int viewLeft, int maxDrag) {
            return ((float) viewLeft) / maxDrag;
        }

        @Override
        public int clampViewPosition(int left, int maxDrag) {
            return Math.max(0, Math.min(left, maxDrag));
        }

        @Override
        public void enableEdgeTrackingOn(ViewDragHelper dragHelper) {
            dragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_LEFT);
        }
    }

    static class RightHelper implements Helper {

        @Override
        public int getLeftAfterFling(float flingVelocity, int maxDrag) {
            return flingVelocity > 0 ? 0 : -maxDrag;
        }

        @Override
        public int getLeftToSettle(float dragProgress, int maxDrag) {
            return dragProgress > 0.5f ? -maxDrag : 0;
        }

        @Override
        public int getRootLeft(float dragProgress, int maxDrag) {
            return (int) -(dragProgress * maxDrag);
        }

        @Override
        public float getDragProgress(int viewLeft, int maxDrag) {
            return ((float) Math.abs(viewLeft)) / maxDrag;
        }

        @Override
        public int clampViewPosition(int left, int maxDrag) {
            return Math.max(-maxDrag, Math.min(left, 0));
        }

        @Override
        public void enableEdgeTrackingOn(ViewDragHelper dragHelper) {
            dragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_RIGHT);
        }
    }

}

File: library/src/main/java/com/yarolegovich/slidingrootnav/SlideGravity.kt
package com.yarolegovich.slidingrootnav

import androidx.customview.widget.ViewDragHelper

/**
 * Created by yarolegovich on 25.03.2017.
 */

enum class SlideGravity {

    LEFT {
        override fun createHelper(): Helper {
            return LeftHelper()
        }
    },
    RIGHT {
        override fun createHelper(): Helper {
            return RightHelper()
        }
    };

    internal abstract fun createHelper(): Helper

    internal interface Helper {

        fun getLeftAfterFling(flingVelocity: Float, maxDrag: Int): Int

        fun getLeftToSettle(dragProgress: Float, maxDrag: Int): Int

        fun getRootLeft(dragProgress: Float, maxDrag: Int): Int

        fun getDragProgress(viewLeft: Int, maxDrag: Int): Float

        fun clampViewPosition(left: Int, maxDrag: Int): Int

        fun enableEdgeTrackingOn(dragHelper: ViewDragHelper)
    }

    internal class LeftHelper : Helper {

        override fun getLeftAfterFling(flingVelocity: Float, maxDrag: Int): Int {
            return if (flingVelocity > 0) maxDrag else 0
        }

        override fun getLeftToSettle(dragProgress: Float, maxDrag: Int): Int {
            return if (dragProgress > 0.5f) maxDrag else 0
        }

        override fun getRootLeft(dragProgress: Float, maxDrag: Int): Int {
            return (dragProgress * maxDrag).toInt()
        }

        override fun getDragProgress(viewLeft: Int, maxDrag: Int): Float {
            return viewLeft.toFloat() / maxDrag
        }

        override fun clampViewPosition(left: Int, maxDrag: Int): Int {
            return Math.max(0, Math.min(left, maxDrag))
        }

        override fun enableEdgeTrackingOn(dragHelper: ViewDragHelper) {
            dragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_LEFT)
        }
    }

    internal class RightHelper : Helper {

        override fun getLeftAfterFling(flingVelocity: Float, maxDrag: Int): Int {
            return if (flingVelocity > 0) 0 else -maxDrag
        }

        override fun getLeftToSettle(dragProgress: Float, maxDrag: Int): Int {
            return if (dragProgress > 0.5f) -maxDrag else 0
        }

        override fun getRootLeft(dragProgress: Float, maxDrag: Int): Int {
            return (-(dragProgress * maxDrag)).toInt()
        }

        override fun getDragProgress(viewLeft: Int, maxDrag: Int): Float {
            return Math.abs(viewLeft).toFloat() / maxDrag
        }

        override fun clampViewPosition(left: Int, maxDrag: Int): Int {
            return Math.max(-maxDrag, Math.min(left, 0))
        }

        override fun enableEdgeTrackingOn(dragHelper: ViewDragHelper) {
            dragHelper.setEdgeTrackingEnabled(ViewDragHelper.EDGE_RIGHT)
        }
    }

}
